nemo_rl/distributed/worker_groups.py

```diff
@@ -487,109 +480,196 @@ def run_all_workers_multiple_data(
         futures = []
 
-        respect_tied_workers = only_on in {"tied_leader", "all_tied_workers"}
+        if run_rank_0_only_axes is None:
+            run_rank_0_only_axes = []
+
+        for worker_idx, worker in enumerate(self.workers):
+            worker_coords = {}
+            for axis in self.sharding_annotations.names:
+                # For this worker, find its position in each axis
+                for i in range(self.sharding_annotations.get_axis_size(axis)):
+                    ranks = self.sharding_annotations.get_ranks(**{axis: i})
+                    if isinstance(ranks, int):
+                        if ranks == worker_idx:
+                            worker_coords[axis] = i
+                            break
+                    elif worker_idx in ranks.layout.flatten():
+                        worker_coords[axis] = i
+                        break
+
+            # Determine if this worker should receive data
+            should_run = True
+            for axis in self.sharding_annotations.names:
+                if axis not in worker_coords:
+                    continue
+                if axis in run_rank_0_only_axes and worker_coords[axis] != 0:
+                    should_run = False
+                    break
 
-        if only_on == "all":
-            for worker in self.workers:
+            if should_run:
                 method = getattr(worker, method_name)
                 futures.append(method.remote(*args, **kwargs))
-        elif respect_tied_workers:
-            for tied_worker_group in self.tied_workers_groups:
-                if only_on == "all_tied_workers":
-                    # Running on all workers in the non-vllm case
-                    for worker_idx in tied_worker_group:
-                        futures.append(
-                            getattr(self._workers[worker_idx], method_name).remote(
-                                *args, **kwargs
-                            )
-                        )
-                else:
-                    futures.append(
-                        getattr(
-                            self._workers[tied_worker_group[0]], method_name
-                        ).remote(*args, **kwargs)
-                    )
-        else:
-            raise ValueError(f"Invalid value for only_on: {only_on}")
 
         return futures
 
+    def run_all_workers_sharded_data(
+        self,
+        method_name: str,
+        data: Iterable[SlicedDataDict],  # arbitrary nested iterables of SlicedDataDicts
+        in_sharded_axes: List[str] | None = None,
+        replicate_on_axes: List[str] | None = None,
+        output_is_replicated: List[str] | None = None,
+        make_dummy_calls_to_free_axes: bool = False,
+        common_kwargs: Optional[Dict[str, Any]] = None,
+    ):
+        """Run a method on all workers in parallel with sharded data.
+
+        All axes provided in in_sharded_axes will be replicated on replicate_on_axes. For axes not provided in either,
```

Comment:
I am not sure if "All axes provided in in_sharded_axes will be replicated on replicate_on_axes" is very clear?
Maybe:

```
Axes in in_sharded_axes: Data is split across workers along these dimensions
Axes in replicate_on_axes: Data is replicated to all workers along these dimensions
Free axes (axes not in either list): Data is only sent to workers at index 0 of these axes
```

Reply:
c477048b43fba01c9deb632b957a2d1c53b0f9c7